net/dns/manager_windows_test.go

```diff
@@ -384,3 +518,97 @@ func (trk *gpNotificationTracker) Close() error {
 	trk.event = 0
 	return nil
 }
+
+type regKeyWatcher struct {
+	keyLocal registry.Key
+	keyGP    registry.Key
+	evtLocal windows.Handle
+	evtGP    windows.Handle
+}
+
+func newRegKeyWatcher() (*regKeyWatcher, error) {
+	var err error
+
+	keyLocal, _, err := registry.CreateKey(registry.LOCAL_MACHINE, nrptBaseLocal, registry.READ)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err != nil {
+			keyLocal.Close()
+		}
+	}()
+
+	// Monitor dnsBaseGP instead of nrptBaseGP, since the latter will be
+	// repeatedly created and destroyed throughout the course of the test.
+	keyGP, _, err := registry.CreateKey(registry.LOCAL_MACHINE, dnsBaseGP, registry.READ)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err != nil {
+			keyGP.Close()
+		}
+	}()
+
+	evtLocal, err := windows.CreateEvent(nil, 0, 0, nil)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err != nil {
+			windows.CloseHandle(evtLocal)
+		}
+	}()
+
+	evtGP, err := windows.CreateEvent(nil, 0, 0, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	result := &regKeyWatcher{
+		keyLocal: keyLocal,
+		keyGP:    keyGP,
+		evtLocal: evtLocal,
+		evtGP:    evtGP,
+	}
+
+	return result, nil
+}
+
+func (rw *regKeyWatcher) watch() error {
+	// We can make these waits thread-agnostic because the tests that use this code must already run on Windows 10+
+	err := windows.RegNotifyChangeKeyValue(windows.Handle(rw.keyLocal), true,
+		windows.REG_NOTIFY_CHANGE_NAME|windows.REG_NOTIFY_THREAD_AGNOSTIC, rw.evtLocal, true)
+	if err != nil {
+		return err
+	}
+
+	return windows.RegNotifyChangeKeyValue(windows.Handle(rw.keyGP), true,
+		windows.REG_NOTIFY_CHANGE_NAME|windows.REG_NOTIFY_THREAD_AGNOSTIC, rw.evtGP, true)
+}
+
+func (rw *regKeyWatcher) wait() error {
+	handles := []windows.Handle{
+		rw.evtLocal,
+		rw.evtGP,
+	}
+
+	waitCode, err := windows.WaitForMultipleObjects(
```

Comment:
would it be good to add some sort of timeout (ideally derived from a context but I'm not picky)

Reply:
Using a context would add a lot more complexity here (due to wiring in its done channel), which I'd take care of if this wasn't test code. In this case I'll just set the timeout directly.